blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ElasticTokensIO.java

```diff
@@ -0,0 +1,291 @@
+            throw new IllegalArgumentException("trying to insert token discovery without a token");
+        }
+
+        return client.prepareIndex(ELASTICSEARCH_TOKEN_INDEX_NAME_WRITE, ES_DOCUMENT_TYPE)
+                .setId(token.getId())
+                .setSource(createSourceContent(token))
+                .setCreate(true)
+                .setRouting(token.getLocator().getTenantId());
+    }
+
+    @Override
+    public List<MetricName> getMetricNames(String tenantId, String query) throws Exception {
+
+        return searchESByIndexes(tenantId, query, getIndexesToSearch());
+    }
+
+    /**
+     * This method queries elastic search for a given glob query and returns list of {@link MetricName}'s.
+     *
+     * @param tenantId
+     * @param query is glob
+     * @param indexes
+     * @return
+     */
+    private List<MetricName> searchESByIndexes(String tenantId, String query, String[] indexes) {
+
+        if (StringUtils.isEmpty(query)) return new ArrayList<>();
+
+        QueryBuilder bqb = buildESQuery(tenantId, query);
+
+        SearchResponse response;
+        Timer.Context timerCtx = esMetricNamesQueryTimer.time();
+        try {
+            response = client.prepareSearch(indexes)
+                             .setRouting(tenantId)
+                             .setSize(AbstractElasticIO.MAX_RESULT_LIMIT)
+                             .setVersion(true)
+                             .setQuery(bqb)
+                             .execute()
+                             .actionGet();
+        } finally {
+            timerCtx.stop();
+        }
+
+        return Arrays.stream(response.getHits().getHits())
+                     .map(this::convertHitToMetricNameResult)
+                     .collect(toList());
+    }
+
+    /**
+     * Builds ES query to grab tokens corresponding to the given query glob.
+     * For a given query foo.bar.*, we would like to grab all the tokens with
+     * parent as foo.bar
+     *
+     * Sample ES query for a query glob = foo.bar.*:
+     *
+     *      "query": {
+     *          "bool" : {
+     *              "must" : [
+     *                  { "term": {  "tenantId": "<tenantId>" }},
+     *                  { "term": {  "parent": "foo.bar" }}
+     *              ]
+     *          }
+     *      }
+     *
+     * @param tenantId
+     * @param query
+     * @return
+     */
+    private BoolQueryBuilder buildESQuery(String tenantId, String query) {
+
+        String[] queryTokens = query.split(Locator.metricTokenSeparatorRegex);
+        String lastToken = queryTokens[queryTokens.length - 1];
+
+        BoolQueryBuilder bqb = boolQuery();
+
+        /**
+         * Builds parent part of the query for the given input query glob tokens.
+         * For a given query foo.bar.*, parent part is foo.bar
+         *
+         * For example:
+         *
+         *  For query = foo.bar.*
+         *          { "term": {  "parent": "foo.bar" }}
+         *
+         *  For query = foo.*.*
```

Comment:
If guess we have to. Someone might choose * from each drop down and plot all metrics in one graph and we support that :)